src/pybind/mgr/mgr_util.py

```diff
@@ -176,3 +178,109 @@ def verify_tls_files(cert_fname, pkey_fname):
         logger.warning(
             'Private key {} and certificate {} do not match up: {}'.format(
                 pkey_fname, cert_fname, str(e)))
+
+def get_most_recent_rate(rates):
+    """ Get most recent rate from rates
+
+    :param rates: The derivative between all time series data points
+    :type rates: list[tuple[float, float]]
+
+    :return: The last derivative or 0.0 if none exists
+    :rtype: float
+
+    >>> get_most_recent_rate(None)
+
+    >>> get_most_recent_rate([(1, -2.0)])
+    -2.0
+    >>> get_most_recent_rate([(1, 2.0), (2, 1.5), (3, 5.0)])
+    5.0
+    """
+    if rates is None or len(rates) == 0:
+        return None
+    return rates[-1][1]
+
+def get_time_series_rates(data):
+    """ Rates from time series data
+
+    :param data: Time series data where the time is provided in seconds
+    :type data: list[tuple[float, float]]
+
+    :return: The derivative between all time series data points
+    :rtype: list[tuple[float, float]]
+
+    >>> logger.debug = lambda x: print(x)
+    >>> get_time_series_rates([])
+
+    >>> get_time_series_rates([[0, 1], [1, 3]])
+    [(1, 2.0)]
+    >>> get_time_series_rates([[0, 2], [0, 3], [0, 1], [1, 2], [1, 3]])
+    Invalid time series samples [[0, 2], [0, 3], [1, 2]] in [[0, 2], [0, 3], [0, 1], [1, 2], [1, 3]]
+    [(1, 2.0)]
+    >>> get_time_series_rates([[1, 1], [2, 3], [4, 11], [5, 16], [6, 22]])
+    [(2, 2.0), (4, 4.0), (5, 5.0), (6, 6.0)]
+    """
+    data = _filter_time_series(data)
+    if data is None:
+        return None
+    return [(data2[0], _derivative(data1, data2)) for data1, data2 in
+            pairwise(data)]
+
+def _filter_time_series(data):
+    """ Filters invalid samples from time series data
+    It also enforces the list to contain at least two samples.
+
+    All filtered values will be shown in the debug log. If values were filtered it's a bug in the
+    time series data collector, please report it.
+
+    :param data: Time series data where the time is provided in seconds
```

Comment:
Can you specify which value of the tuple corresponds to the timestamp and the value?

Reply:
done